libs/message-coder/src/message_coder.ts

```diff
@@ -0,0 +1,123 @@
+import { ok } from '@votingworks/types';
+import { Buffer } from 'buffer';
+import { BaseCoder } from './base_coder';
+import { LiteralCoder } from './literal_coder';
+import { PaddingCoder } from './padding_coder';
+import { BitOffset, Coder, DecodeResult, EncodeResult } from './types';
+
+/**
+ * Parts of a message coder.
+ */
+type MessageCoderParts<T> = {
+  [K in keyof T]: Coder<T[K]> | LiteralCoder | PaddingCoder;
+};
+type PropsMatching<T, V> = {
+  [P in keyof T]: T[P] extends V ? P : never;
+}[keyof T];
+type FilterOut<T, V> = Omit<T, PropsMatching<T, V>>;
+
+/**
+ * Gets the type encoded by a coder.
+ */
+export type CoderType<T> = T extends Coder<infer U> ? U : never;
+
+/**
+ * Gets the object type encoded by a message coder's parts.
+ */
+type CoderPartsOperand<T> = T extends MessageCoderParts<infer U> ? U : never;
+
+/**
+ * Gets the type of a coder from its parts. Excludes any coders that don't
+ * contribute to the JS object.
+ */
+type CoderFromParts<T> = Coder<
+  CoderPartsOperand<FilterOut<T, LiteralCoder | PaddingCoder>>
+>;
```

Comment:
This is the utility type that does the filtering of literals and padding properties.